### ISSUE TITLE
MySQL server has gone away

### ISSUE THREAD

**Issue by moneroexamples** [moneroexamples/openmonero#121]
```
# ERR: SQLException in /home/mwo2/openmonero/src/MySqlAccounts.cpp(select) on line 258
# ERR: MySQL server has gone away
# ERR: SQLException in /home/mwo2/openmonero/src/MySqlAccounts.cpp(insert) on line 285
# ERR: MySQL server has gone away
```
Sometimes it happens. Don't know yet why.

**Comment by italocoin-project** [comment]
This could be fixed with force connect check this maybe can help, i haven't encountered this issue maybe because of this

https://github.com/italocoin-project/bittube-light-wallet/commit/d34cf46abbf59feec7415aa17bcaa94797973b92

**Comment by moneroexamples** [comment]
Thx. Will check that.

**Comment by moneroexamples** [comment]
Thanks. I checked the code and this may work, and will reconnect you. But the problem I see is that the query data will be lost. The code that is calling these queries will not try to re-call it again. So you may end up with missing outputs or inputs in your database after reconnecting in `catch` part.

I think for reconnection, the better place would be before the query is executed, i.e., where `conn->check_if_connected();`. So maybe this could be modified, so that it reconnects.

There can also be other reasons why the connection is failing, e.g., default settings in mysql/mariadb being inadequate.

- https://matomo.org/faq/troubleshooting/faq_183/
- https://stackoverflow.com/questions/7942154/mysql-error-2006-mysql-server-has-gone-away

I will try to investigate why its disconnection. Maybe mysql/mariadb logs will some info about that.

**Comment by italocoin-project** [comment]
No problem, i see, this could bring up some problems if it would miss outputs or inputs from mysql DB, but then, if the input gets missed, on a rescan it should be reeaded, because the rescan will find the imputs/outputs from blockchain's db, isn't it? let me know if you find a better way to deal with timeouts, thanks :)

**Comment by moneroexamples** [comment]
no problem and yes, would have to rescan to find missing txs/outputs/inputs.

I just setup regular mariadb on my linux box, instead of using docker, and will try to get some logs out of it.

**Comment by moneroexamples** [comment]
I added `ReconnectOption` to mysql, so now it will reconnect when connection drops.

https://github.com/moneroexamples/openmonero/blob/9e3b72fb75167a535986c79894d0b272250cfed7/src/db/MySqlAccounts.cpp#L228-L229

There still can be a problem, but at least connection should not be dropped permanently. If I manually kill the connection in mariadb I get this, but then it reconnects:

```
# ERR: SQLException in /home/mwo2/openmonero/src/db/MySqlAccounts.cpp(select) on line 267                                                                                                                                                                                   
# ERR: Lost connection to MySQL server during query  
```

In the current devel branch I also added Ctrl+C support to greacfully stop openmonero, and the logging got extended.

**Comment by italocoin-project** [comment]
You think this option is better then force connect that i've mentioned above?
nice work with Ctrl+C support, great job man.

I have another issue: `MDB_READERS_FULL issue`
basicaly when i get MANY connections i get the `MDB_READERS_FULL error` under heavy load the db leaks readers or something. Did you encountered this problem?

**Comment by italocoin-project** [comment]
> How many is MANY?
Send more than 126 different login request to the mywallet and you will get the issue, let me know 

yes, now it actually closes properly, what if it gets killed?

**Comment by moneroexamples** [comment]
> 126 different login request to the mywallet 

I see. Probably will need to rethink how lmdb is accessed in openmonero. At the moment, there will be 126 search tx threads accessing the lmdb at the same time. So maybe will have to implement some pooling when only few threads access lmdb at the same time, or move to using rpc to deamon instead of accessing lmdb directly. 

> what if it gets killed?

Thank it will be like before, all connections, threads will be terminated immedietly. I think it's better to let the all threads, mysql and lmdb close properly when openmonero stops.

**Comment by italocoin-project** [comment]
> or move to using rpc to deamon instead of accessing lmdb directly

rpc to daemon could solve the problem for good.

> maybe will have to implement some pooling when only few threads access lmdb at the same time

It means that someone will have to wait until the threads gets freed again, i think we should test them both

**Comment by moneroexamples** [comment]
rpc to daemon should also be much easier to implement, as the second possible solution would require major changes to architecture of openmonero. But rpc will also not be simultaneous, as is needs also synchronization.

**Comment by italocoin-project** [comment]
> rpc to daemon should also be much easier to implement, as the second possible solution would require major changes to architecture of openmonero.

I agree 

>  But rpc will also not be simultaneous, as is needs also synchronization.

Thats also true, but i think the best solution is the rpc to daemon, how is mymonero handling with many connections? I assume that they have over 126 consecutive login requests

**Comment by italocoin-project** [comment]
> so I think rpc also would be easiest.

Yes, that would be the best option, do you plan to implement RPC to DAEMON soon? And about more instances, that would be a good idea too, a system to sync all mysql db on all instances would solve the panic in users if they would see that the wallet syncs from 0 again

**Comment by moneroexamples** [comment]
> Yes, that would be the best option, do you plan to implement RPC to DAEMON soon? 

Yes, though "soon" in monero world does not mean anything:-) But yes, this and mysql-thing will be the issues I will be focusing the most at present. When complete? No timeframe on this. Hopefully "soon":-)

Just spin up new branch for that: https://github.com/moneroexamples/openmonero/tree/new_rpc So all new changes will be there for experimentation.

**Comment by italocoin-project** [comment]
Nice, i will merge it tomorrow morning and test out the branch, i will let you know if i face issues, thanks :)

**Comment by moneroexamples** [comment]
Not sure I will have there anything new there for tomorrow, but this is the branch when the rpc changes will be taking place.

No problem. Good you can provide extra info and much needed testing for openmonero. Thanks.

**Comment by italocoin-project** [comment]
Thank you for your hard work, i really like your projects and work, you can count on my help for testing out, i use openmonero for my project to, thanks for that :)
